### ISSUE TITLE
expl3 "typesetting" module 

### ISSUE THREAD

**Issue by wspr** [latex3/latex3#575]
Joseph brings up a good point in a recent discussion. There are a number of typesetting concepts that are not yet supported in expl3 and for doing simple things like framing boxes we need things like rules, kerns, etc.

It's probably time to think about a "typesetting" module (except with a more specific name) that can handle these concepts.

There are possibly a couple commands in l3box that should be taken out from that module and moved into this new hypothetical one.

**Comment by car222222** [comment]
True, but it also needs a “scripting language” for declaring simple layout and formatting.  A language whose use does not demand knowledge of the more obscure details of either TeX’s typesetting or L3PL’s syntax and semantics.

**Comment by cmplstofB** [comment]
Would you like to consider about something to underline, which LaTeX2e is not good at?

**Comment by wspr** [comment]
Not at the “layer of the onion”. Both Frank and you (I think) have presented at different times the need for design-level interfaces, but the need here is literally just “if I am using expl3, how am I supposed to input a kern, or a rule, or a phantom, …”.

I.e., definitely not user or designer level, still quick close to the metal in terms of primitives but a clean interface with the other expl3 ideas.

**Comment by car222222** [comment]
Agreed.  I just thought it was a good opportunity to publicise this onion-level.

**Comment by tail-reversion** [comment]
Is there any reason why there’s no expl3 equivalent of `\mskip`? It seems odd that all the functions for creating and manipulating muskips exist yet there’s no way to use muskips.

**Comment by josephwright** [comment]
@tail-reversion There are several gaps beyond `\tex_...:D` in the 'typesetting' area: a bigger case than `\mskip` would be `\kern`. We still do not have a position on the detail of interfaces here: we do know there's a gap!

**Comment by tail-reversion** [comment]
I’ve been thinking, advanced typesetting might be eased by having strict control over the order of items inserted into the various lists… Perhaps, instead of directly inserting things like kerns, skips, and penalties, they could be specified during box insertion.

For example, to kern two items, there could be `\box_use_with_kern:NNN`, where `#1` and `#2` are boxes, and `#3` is a kern to insert between them.

**Comment by tail-reversion** [comment]
After some thought, I realize that my previous comment doesn’t work very well—at least, not in the general case.

It is tempting to request the simple interface of public expl3 equivalents of the `\tex_...:D` csnames, but I recognize that that might cause issues, down the road.
